src/librustc_resolve/lib.rs

```diff
@@ -3186,23 +3186,42 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                 let enum_candidates =
                     this.lookup_import_candidates(ident.name, ns, is_enum_variant);
                 let mut enum_candidates = enum_candidates.iter()
-                    .map(|suggestion| import_candidate_to_paths(&suggestion)).collect::<Vec<_>>();
+                    .map(|suggestion| {
+                        import_candidate_to_enum_paths(&suggestion)
+                    }).collect::<Vec<_>>();
                 enum_candidates.sort();
-                for (sp, variant_path, enum_path) in enum_candidates {
-                    if sp.is_dummy() {
-                        let msg = format!("there is an enum variant `{}`, \
-                                        try using `{}`?",
-                                        variant_path,
-                                        enum_path);
-                        err.help(&msg);
+
+                if !enum_candidates.is_empty() {
+                    // contextualize for E0412 "cannot find type", but don't belabor the point
+                    // (that it's a variant) for E0573 "expected type, found variant"
+                    let preamble = if def.is_none() {
+                        let others = match enum_candidates.len() {
+                            1 => String::new(),
+                            2 => " and 1 other".to_owned(),
+                            n => format!(" and {} others", n)
+                        };
+                        format!("there is an enum variant `{}`{}; ",
+                                enum_candidates[0].0, others)
                     } else {
-                        err.span_suggestion_with_applicability(
-                            span,
-                            "you can try using the variant's enum",
-                            enum_path,
-                            Applicability::MachineApplicable,
-                        );
-                    }
+                        String::new()
+                    };
+                    let msg = format!("{}try using the variant's enum", preamble);
+
+                    err.span_suggestions_with_applicability(
+                        span,
+                        &msg,
+                        enum_candidates.into_iter()
+                            .map(|(_variant_path, enum_ty_path)| enum_ty_path)
+                            // variants reëxported in prelude doesn't mean `prelude::v1` is the
+                            // type name! FIXME: is there a more principled way to do this that
+                            // would work for other reëxports?
+                            .filter(|enum_ty_path| enum_ty_path != "std::prelude::v1")
+                            // also say `Option` rather than `std::prelude::v1::Option`
+                            .map(|enum_ty_path| {
+                                enum_ty_path.trim_left_matches("std::prelude::v1::").to_owned()
```

Comment:
I would normally suggest using [`item_path_str`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.item_path_str) but you don't have access to `TyCtxt` here.

Not sure if this will work, but there's a [`prelude`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_resolve/struct.Resolver.html#structfield.prelude) field on `Resolver` and a [`module_to_string`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_resolve/fn.module_to_string.html) function - perhaps that will get you the equivalent of `"std::prelude::v1::"` without hardcoding it?

If that doesn't work, I've found [one other location where this is hardcoded](https://github.com/rust-lang/rust/blob/451987d86c89b38ddd8c4c124f1b9b6d4ded6983/src/librustc_typeck/check/demand.rs#L126) - perhaps make a function on `Session` that returns a `&'static str` that you can call in this code and that code to stop the duplication - not ideal but it'd be a little better?

Reply:
There are also a handful of tests that have `std::prelude::v1` in the output, so I guess we could leave that in there too.

Reply:
I agree that the duplicated hardcoded path is bad, but is it OK to defer that to a future PR? (I'd want to decide which de-duplication strategy is best while cleaning up all the other instances of `std::prelude` in tests.) I filed a dedicated issue so that we don't forget: #56861

Reply:
I'm don't mind having this deferred for now, looks like the previous instance of this was added in #43178. I'd like to see #56861 resolved soon though.